drv/opentitan-aontimer/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+
+use core::ptr::slice_from_raw_parts_mut;
+
+const WDOG_REGWEN_IDX: usize = 4;
+const WDOG_CTRL_IDX: usize = 5;
+const WDOG_BARK_THOLD_IDX: usize = 6;
+const WDOG_BITE_THOLD_IDX: usize = 7;
+const WDOG_COUNT_IDX: usize = 8;
+const INTR_STATE_IDX: usize = 9;
+
+// The OpenTitan Always-On Timer is a counter that increments. If the counter
+// reaches a configurable threshold, it will emit an interrupt signal, called the
+// "bark". If the counter reaches a second configurable threshold, it will reset
+// the system, called the "bite".
+// Enabling the AON Timer before setting a bite threshold causes the system to
+// immediately reset.
+pub struct AonTimer {
+    base_addr: *mut [u32],
+    clock_freq_hz: u32,
+    pub bark_cb: Option<fn()>,
+}
+
+impl AonTimer {
+    // Initialize a new watchdog driver device with the specified base address,
+    // clock frequency (in Hz), and the bark/bite thresholds (both in ms).
+    // If the bark threshold > bite threshold, the bite threshold is set to
+    // the bark threshold.
+    pub fn new(
+        base_addr: u32,
```

Comment:
The way the code is currently organized where this driver is a library, probably having base address as an input is fine. But, the base address isn't really something that is task configurable. It is fixed for a chip. So, is there any benefit of allowing base address to be taken in as an input?

Reply:
You're right. At the time it was written, it was meant to be a more generic interface for watchdogs, but that changed as time went on.

Reply:
I was actually thinking the opposite.  With the driver being in its own crate, I think it should be agnostic to the chip memory map.  So the task using the driver will tell the crate what addresses to operate at.

Reply:
So, it depends on what the driver is intended for. Let's say the driver is meant only for opentitan earlgrey chip. Then, passing in a base address is unnecessary. However, if the driver is meant to be used by (let's say for example) multiple opentitan chips with different memory layouts, then yes, accepting base address as an input here makes sense. However, in either case, I don't think the address should be coming in from a "task". Task should not have to worry about any chip-level details.

Reply:
Once there are[ build consts for peripherals](https://rivosinc.atlassian.net/browse/SW-431), the task can just pass the const though to the driver crate.  It might also be possible for the driver crate to directly access the const.

Reply:
We can keep this as is for now. But, we should revisit this in the future. I think accepting this address from tasks also opens up the condition that task might pass in an incorrect address and so needs to be handled correctly. Instead if it is a chip attribute, it should just be something that is embedded directly within the driver.